knodle/trainer/crossweigh_weighing/data_splitting_by_rules.py

```diff
@@ -0,0 +1,289 @@
+
+
+def get_signature_sample_ids(rule_matches_z: np.ndarray) -> (Dict, Dict):
+    """
+    This function calculates the signature for each sample (e.g. if rules with ids 1, 5, 7 matched in a sample, a sample
+    signature id is 1_5_7) and id for each signature.
+    :param rule_matches_z: matrix of rule matches (num_samples x num_rules)
+    :return: two dictionaries:
+        - {signature: signature id}
+        - {signature id: [sample_id1, sample_id2, ...]}
+    """
+
+    signature2id, signature_id2samples = {}, {}
+
+    if isinstance(rule_matches_z, sp.csr_matrix):
+        samples_id_rules_dict = {key: [] for key in range(rule_matches_z.shape[0])}
+        for row_id, col_id in zip(*rule_matches_z.nonzero()):
+            samples_id_rules_dict[row_id].append(col_id)
+
+    else:
+        samples_id_rules_dict = dict((i, set()) for i in range(0, rule_matches_z.shape[0]))
+        for row_id, row in enumerate(rule_matches_z):
+            samples_id_rules_dict[row_id] = np.where(row == 1)[0].tolist()
+
+    for sample_id, rules in samples_id_rules_dict.items():
+        signature = "_".join(map(str, sorted(list(rules))))
+
+        if signature in signature2id:
+            signature_id = signature2id.get(signature)
+        else:
+            signature_id = len(signature2id)
+            signature2id[signature] = signature_id
+
+        if signature_id in signature_id2samples:
+            signature_id2samples[signature_id].append(sample_id)
+        else:
+            signature_id2samples[signature_id] = [sample_id]
+
+    return signature2id, signature_id2samples
+
+
+def compose_train_n_test_datasets(
+        data_features: np.ndarray, rule_id2samples_ids: Dict, labels: np.ndarray, folds: int,
+        partitions: int, other_class_id: int = None
+) -> (List, List):
+    """
+    This function creates train and test datasets for k-folds cross-validation.
+
+    :param data_features: encoded data samples (num_samples x num_features):
+    :param rule_id2samples_ids: {rule_id: [sample_id1, sample_id2, ...]}. Rule in this context means anything basing on
+    which splitting is performed (matched rule, sample signature, ...).
+    :param labels: array of labels
+    :param folds: number of folds the data instances are to be slitted into in each partition
+    :param partitions: number of partitions that are to be performed
+    :param other_class_id: if you don't want to include the negative samples (the ones that belong to the other class)
+     to the test set, but only to the training set, you can pass the id of other class should be
+
+    :return: list of train sets and list of corresponding test (hold-out) sets
+    """
+
+    # calculate ids of all samples that belong to the 'other_class'
+    other_sample_ids = np.where(labels[:, other_class_id] == 1)[0].tolist() if other_class_id else None
+    # make a list of rule ids to shuffle them later
+    rule_ids = [rule_id for rule_id in range(0, len(rule_id2samples_ids))]
```

Comment:
*refactor int ids part 2* 
Only this line assumes that rule_ids are actually in range [0, len(rule_id2samples_ids)]. As far as I understand, shuffling actual keys here (even if strings) does not hurt anywhere later: shuffling would work fine, as well as the picking the fold rule_ids in calculate_rules_indices `calculate_rules_indices` and extracting the actual samples in `get_samples_labels_idx_by_rule_id`
```suggestion
    rule_ids = list(rule_id2samples_ids.keys())
```